internal/cloud/awscloud/secure-instance.go

```diff
@@ -0,0 +1,370 @@
+		InstanceIds: []*string{
+			si.Instance.InstanceId,
+		},
+	})
+	if err != nil {
+		return err
+	}
+	_, err = a.ec2.DeleteSecurityGroup(&ec2.DeleteSecurityGroupInput{
+		GroupId: aws.String(si.SGID),
+	})
+	if err != nil {
+		return err
+	}
+
+	_, err = a.ec2.DeleteLaunchTemplate(&ec2.DeleteLaunchTemplateInput{
+		LaunchTemplateId: aws.String(si.LTID),
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (a *AWS) createOrReplaceSG(hostInstanceID, hostIP, vpcID string) (string, error) {
+	sgName := fmt.Sprintf("SG for %s (%s)", hostInstanceID, hostIP)
+	descrSGOutput, err := a.ec2.DescribeSecurityGroups(&ec2.DescribeSecurityGroupsInput{
+		GroupNames: []*string{
+			aws.String(sgName),
+		},
+	})
+	if descrSGOutput == nil {
+		return "", fmt.Errorf("Not able to describe security groups")
+	}
+	if len(descrSGOutput.SecurityGroups) == 1 {
+		_, err := a.ec2.DeleteSecurityGroup(&ec2.DeleteSecurityGroupInput{
+			GroupId: descrSGOutput.SecurityGroups[0].GroupId,
+		})
+		if err != nil {
+			return "", err
+		}
+	}
+	if err != nil {
+		return "", err
+	}
+
+	cSGOutput, err := a.ec2.CreateSecurityGroup(&ec2.CreateSecurityGroupInput{
+		Description: aws.String(sgName),
+		GroupName:   aws.String(sgName),
+		VpcId:       aws.String(vpcID),
+	})
+	if err != nil {
+		return "", err
+	}
+	sgID := *cSGOutput.GroupId
+
+	sgIngressOutput, err := a.ec2.AuthorizeSecurityGroupIngress(&ec2.AuthorizeSecurityGroupIngressInput{
+		GroupId: aws.String(sgID),
+		IpPermissions: []*ec2.IpPermission{
+			&ec2.IpPermission{
+				IpProtocol: aws.String(ec2.ProtocolTcp),
+				FromPort:   aws.Int64(1),
+				ToPort:     aws.Int64(65535),
+				IpRanges: []*ec2.IpRange{
+					&ec2.IpRange{
+						CidrIp: aws.String(fmt.Sprintf("%s/32", hostIP)),
+					},
+				},
+			},
+		},
+	})
+	if err != nil {
+		return sgID, err
+	}
+	if *sgIngressOutput.Return {
+		return sgID, fmt.Errorf("Unable to attach ingress rules to SG")
+	}
+
+	sgEgressOutput, err := a.ec2.AuthorizeSecurityGroupEgress(&ec2.AuthorizeSecurityGroupEgressInput{
+		GroupId: aws.String(sgID),
+		IpPermissions: []*ec2.IpPermission{
+			&ec2.IpPermission{
+				IpProtocol: aws.String(ec2.ProtocolTcp),
+				FromPort:   aws.Int64(1),
+				ToPort:     aws.Int64(65535),
+				IpRanges: []*ec2.IpRange{
+					&ec2.IpRange{
+						CidrIp: aws.String(fmt.Sprintf("%s/32", hostIP)),
+					},
+				},
+			},
+		},
+	})
+	if err != nil {
+		return sgID, err
+	}
+	if *sgEgressOutput.Return {
+		return sgID, fmt.Errorf("Unable to attach egress rules to SG")
+	}
+
+	describeSGOutput, err := a.ec2.DescribeSecurityGroups(&ec2.DescribeSecurityGroupsInput{
+		GroupIds: []*string{
+			aws.String(sgID),
+		},
+	})
+	if err != nil {
+		return sgID, err
+	}
+	if len(describeSGOutput.SecurityGroups) != 1 {
+		return sgID, fmt.Errorf("Expected 1 security group")
+	}
+	if len(describeSGOutput.SecurityGroups[0].IpPermissions) != 1 || len(describeSGOutput.SecurityGroups[0].IpPermissionsEgress) != 2 {
```

Comment:
there's also a default rule allowing all outgoing traffic. Since we're still pulling sources we need it. In future we can remove it when the worker is a full proxy for the sources.

Reply:
Maybe let's add a comment about it? It's not obvious on the first glance.